### PR TITLE
Fix #281639: Better handling of Clefs, Barlines on ossias and cutaway

### PR DESCRIPTION
**Pull request by elerouxx**
Partially Resolves: *https://musescore.org/en/node/281639*
(Brackets out of scope of this PR)

**Fixes visibility of Bar Lines for Ossias or Cutaway scores.**
The EndBarLine of an invisible measure will automatically show if next measure is visible and in the same system.
Span works as expected, no manual insert of bar line needed.
(Barlines can always be manually set to invisible, if desired)

**Fixes showing courtesy clefs on invisible measures**
On a cutaway staff, If a courtesy clef is added to the end of an empty measure, a short staff "trailer" will be created to display the clef. This allows clefs to be displayed on the first visible measure or after an invisible section of a cutaway staff.
  (Clefs must be created manually by adding the clef to a  whole measure, as always).

Before
image

After
image

As always, feedback and review will be appreciated.

*Some additional small fixes:
- Dragging manually-added bar lines won't set the Global span anymore (which was probably the intention of the existing code)
- Measure numbers will show even if measure is invisible or cut away from the system.
- Spacers will always show, even on invisible measures, to avoid losing track of them on a cutaway score.*

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://github.com/musescore/Documentation/blob/master/WorkflowAndGuidelines/CodeGuidelines.md)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [] I created the test (mtest, vtest, script test) to verify the changes I made

### CODE REVIEW

**Comment by elerouxx** [comment]
image
This was the only difference in visual tests (as expected)

**Comment by vpereverzev** [comment]
Looking for rebase

**Comment by elerouxx** [comment]
> Looking for rebase

Done!

File: libmscore/measure.cpp [hunk 1]
```diff
@@ -261,8 +261,16 @@ Measure::Measure(const Measure& m)
 
 void Measure::layoutStaffLines()
       {
-      for (MStaff* ms : _mstaves)
-            ms->lines()->layout();
+      int staffIdx = 0;
+      for (MStaff* ms : _mstaves) { 
+            if (isCutawayClef(staffIdx) && (score()->staff(staffIdx)->cutaway() || !visible(staffIdx)))
```
Comment (Marr11317) [inline libmscore/measure.cpp:266]: there's a function in the `Measure` class for that: `Measure::visible(int staffIdx)`

Comment (Marr11317) [reply libmscore/measure.cpp:266]: Not sure if it's more appropriate though.

File: libmscore/measure.cpp [hunk 2]
```diff
@@ -2588,10 +2596,10 @@ void Measure::scanElements(void* data, void (*func)(void*, Element*), bool all)
 
       int nstaves = score()->nstaves();
       for (int staffIdx = 0; staffIdx < nstaves; ++staffIdx) {
-            if (!all && !(visible(staffIdx) && score()->staff(staffIdx)->show()))
+            if (!all && !score()->staff(staffIdx)->show())
```
Comment (Marr11317) [inline libmscore/measure.cpp:2599]: why did you remove that `!(visible(staffIdx))` check? You left only the score-wise check and removed the measure-wise check.

File: libmscore/segment.cpp [hunk 3]
```diff
@@ -1165,13 +1165,29 @@ void Segment::scanElements(void* data, void (*func)(void*, Element*), bool all)
       {
       for (int track = 0; track < score()->nstaves() * VOICES; ++track) {
             int staffIdx = track/VOICES;
-            if (!all && !(measure()->visible(staffIdx) && score()->staff(staffIdx)->show())) {
+            if (!all && !(/*measure()->visible(staffIdx) &&*/ score()->staff(staffIdx)->show())) {
```
Comment (ypoissant) [inline libmscore/segment.cpp:1168]: Should the commented out code be removed or reinstated?

File: libmscore/measure.cpp (hunk 2 above)
Comment (ypoissant) [inline libmscore/measure.cpp:2599]: What happens if show is called while staffIdx is not visible?

File: libmscore/measure.cpp (hunk 1 above)
Comment (elerouxx) [reply libmscore/measure.cpp:266]: It's being used there. Means: if the measure has only a clef before end barline, and either (the staff is cutaway or the measure is invisible), then go ahead and draw the staff lines. It will later draw only a small length of the lines to hold the 'courtesy' cutaway clef. 
Otherwise there will be no staff lines at all

File: libmscore/measure.cpp (hunk 2 above)
Comment (elerouxx) [reply libmscore/measure.cpp:2599]: I'll check for potential exceptions here, thank you!.

File: libmscore/segment.cpp (hunk 3 above)
Comment (elerouxx) [reply libmscore/segment.cpp:1168]: That would be a check if the measure is marked as visible. I'll double check. Thanks for helping me by reviewing the code.

File: libmscore/measure.cpp (hunk 2 above)
Comment (Marr11317) [reply libmscore/measure.cpp:2599]: there's other elements than spacers and breaks in this for. Maybe include this check somewhere else for the other elements?

Comment (elerouxx) [reply libmscore/measure.cpp:2599]: Sorry, I meant breaks, spacers, measure numbers and mmRangeText, introduced recently by you, but none seems related to the exception introduced some merges ago.

Comment (elerouxx) [reply libmscore/measure.cpp:2599]: No problem, it's just a check if the measure is set as invisible. Not related to the exception I'm chasing. It was introduced before this PR.

File: libmscore/segment.cpp (hunk 3 above)
Comment (elerouxx) [reply libmscore/segment.cpp]: Removed it.

**Comment by elerouxx** [comment]
Updated and ready.
